RFC Stages

jenweber opened the pull request:
[Rendered](https://github.com/jenweber/rfcs/blob/stages/text/0617-rfc-stages.md)

(Author's note - this RFC has been substantially updated to reflect community & core team feedback. Thanks all for your time and patience!)

_Edited 7/28 to update the `Rendered` link._

locks commented on text/0000-rfc-stages.md:
```diff
@@ -0,0 +1,189 @@
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
```
```suggestion
- Start Date: 2020-04-22
- Relevant Team(s): (fill this in with the team(s) to which this RFC applies)
```

locks commented on text/0000-rfc-stages.md:
```diff
@@ -0,0 +1,189 @@
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: (leave this empty)
```
```suggestion
- RFC PR: https://github.com/emberjs/rfcs/pull/617
```

locks commented on text/0000-rfc-stages.md:
```diff
@@ -0,0 +1,189 @@
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
```
```suggestion
- Tracking: (leave this empty)
```

mansona commented on text/0000-rfc-stages.md:
```diff
@@ -0,0 +1,189 @@
+
+The Core Teams' review of readiness for release has historically been an informal process. This RFC formalizes the responsibility by creating a stage for gathering confirmation that all necessary work is complete.
+
+When the work described in an RFC is ready for Release, a PR to make the change to the RFC is added by the RFC champion, and each team is requested as a reviewer. A representative of each team adds a review, and the community is also invited to comment via one week of FCP (final comment period). If a team does not respond to the request, it is assumed that the release may proceed. 
```
How does this "one week limit" sync with the planned change to the learning team's meeting schedule to be one meeting a month? 🤔 I guess this doesn't really need to be something that **must** be discussed at a "formal" meeting, I'm only asking because we tried to keep up with an asynchronous process last time we tried to work on RFC tracking in the Learning team

mansona commented on text/0000-rfc-stages.md:
```diff
@@ -0,0 +1,189 @@
+
+## Drawbacks
+
+Making updates to past RFCs can generate a lot of  GitHub notifications for people who watch the repository. We think that the benefit of knowing when an RFC is available in a stable release outweighs this drawback. Additionally, in the future, if an app displays the frontmatter data, someone could use that as their main source of information. Lastly, the Ember Times does a great job curating RFC ...
```
I don't know if this is the place to leave comments about possible workarounds, but we could limit the change of past RFCs to one single PR so that it only creates a single notification for people watching the repo. I guess it would be a tradeoff between the size of this PR and the number of notifications 🤔

mansona replied:
I don't know if this is the place to leave comments about possible workarounds, but we could limit the change of past RFCs to one single PR so that it only creates a single notification for people watching the repo. I guess it would be a tradeoff between the size of this PR and the number of notifications 🤔

mehulkar commented on text/0000-rfc-stages.md:
```diff
@@ -0,0 +1,189 @@
+| Accepted | "The work to integrate the proposal into Ember's codebases or materials may begin. An ""Accepted"" RFC is not necessary in-progress." | "The technical implementation is complete according to plan outlined in the RFC, and is in harmony with any changes in Ember that have occurred since the RFC was first written." |
+| Prepared for Release | "The relevant core team reviews the work to determine when it can be included in a stable release. At this stage, features or deprecations may be available for use behind a feature flag, in beta, etc. This is also a time to queue up learning materials, for example." | "The relevant core teams coordinate to include the work in a stable release, and/or publish the materials ...
+| Stable | "The work is in a stable release/is published. If it is codebase-related work, it adheres to SemVer from this point onward." | "If it is a codebase change, it is well documented and has clear migration paths. It is consistent with Ember's mental models." |
+| Complete | The feature/resource is recommended for general use. | n/a |
```
I feel like these stages do a good job of capturing status after RFCs are merged, but the "Proposed" step is what needs to be expanded/exploded. As an RFC author, that part is completely opaque and really doesn't motivate me to write more RFCs.

If these were design documents / RFCs for a product, there would be some SLA or some kind of rubric to determine how to move forward with a document.

Another thing to think about is expirations. I think it's totally reasonable to expire an RFC and close it out if it doesn't receive either consensus or updates for N months (or whatever the criteria is). It's incredibly frustrating to see RFCs stay open forever. I imagine that this would reduce maintainer burden as well. PRs and Issues don't go anywhere, so it would be fine to open them back up again or resubmit some other way.

kategengler replied:
I think you are correct in that this RFC is not exactly intended to address those particular problems with RFCs. I do think the `Exploring` stage will help for a subset -- the core teams can quickly indicate that an RFC deserves further effort. 

Recently, at least the framework team, we have made an effort to ensure team notes have been published much more frequently. That said, when an RFC is discussed, we nearly always comment on that RFC, either with requested changes or any action that was decided.

Gaurav0 commented:
I'd like to propose that proposed RFC's, those that have been pull requested but haven't been put into final comment period, have at least 3 stages:

Stage 1: An initial proposal. A pull request has been opened. Community feedback is desired.
Stage 2: A champion has been chosen and is shepherding the proposal through the process.
Stage 3: A champion has determined that the proposal is fit enough that it is appropriate to begin discussions behind the scenes with the core team. If a community member is inclined, they could begin writing an addon or other community solution to solidify the concept.

It is fairly clear when a pull request is open today and when it is put into final comment period. It is often not clear when stage 2 and especially stage 3 as I have described has occurred. Knowing whether a proposal should be considered stage 3 would be very helpful to even the writer and especially to potential implementers.

locks commented on text/0000-rfc-stages.md:
```diff
@@ -0,0 +1,189 @@
+or some semi-obvious variation of it." (When the answer to the latter question
+is "no", then the appropriate response is to close the RFC, not postpone it.)
+
+ As of the writing of this RFC, there are over [70 open pull requests](https://github.com/emberjs/rfcs/pulls) for Ember's RFCs. Having so many open PRs is confusing for everyone but the core team, yet simply closing many of these PRs would not reflect the reality of the wish to bring these good ideas into Ember.
```
```suggestion
 As of the writing of this RFC, there are over [70 open pull requests](https://github.com/emberjs/rfcs/pulls) for Ember's RFCs. Having so many open PRs is confusing for everyone, yet simply closing many of these PRs would not reflect the reality of the wish to bring these good ideas into Ember.
```

MelSumner commented:
@jenweber it's super important to me that the RFC explicitly talks about the recourse for what we do if an RFC has had an experimental trial but significant issues have been discovered. As a core team member, I'd like a clear way to withdraw consensus.

jenweber commented:
Hi everyone! Thanks so much for your thoughtful feedback & ideas. Two themes emerged, and I've made changes to reflect them: 
1. The community wanted more insight into what happens between the time when a PR is opened and merged, so we added an "Exploring" stage to this proposal. 
2. The framework team needed a little more details about how FCP is done and what this would all look like in practice. I added an "FCP" column to the stages table, and a case study for how this would have worked for Tracked Properties, if we had this process to follow.

Now that these changes are in place, I think this is ready for another review from anyone interested.

Special thanks to @pzuraq for pairing on the latest round of edits and helping to fill in the Case Study info!

jenweber commented:
P.S. I'm an alumni of the framework team and can't come to many meetings these days, so @pzuraq has volunteered to help champion this RFC. I am available to help out as-needed!

kategengler commented on text/0000-rfc-stages.md:
```diff
@@ -0,0 +1,189 @@
+| Stage| Description| Criteria to advance to next step |
+| -----| -----------|----------------------------------| 
+| Proposed | A pull request made to the RFC repository which is open for public comment and core team evaluation. | "The RFC has been in Final Comment Period, and the relevant core team has consensus to merge it, thus accepting the proposal." |
+| Accepted | "The work to integrate the proposal into Ember's codebases or materials may begin. An ""Accepted"" RFC is not necessary in-progress." | "The technical implementation is complete according to plan outlined in the RFC, and is in harmony with any changes in Ember that have occurred since the RFC was first written." |
```
Realistically, an RFC can be implemented even before it is opened. It's not clear that "in-progress" is a stage rather than a status that can be concurrent with any of the stages before "Released". This RFC was just updated with more detail, the current implementation status should be clear in the PR for "Ready for Release" for the RFC.

mansona replied:
@kategengler yea you make a good point. I just re-read the RFC with the most recent updates and I'm happy that the "status" of the RFC can be represented effectively 👍

mehulkar commented on text/0617-rfc-stages.md:
```diff
@@ -0,0 +1,321 @@
+| [2 - Accepted](#Accepted) | A fully specified RFC. Waiting for or in the process of implementation. | yes |
+| [3 - Ready for Release](#Ready-for-Release) | The implementation of the RFC is complete, including learning materials. | yes |
+| [4 - Released](#Released) | The work is published. If it is codebase-related work, it is in a stable version of the relevant package(s). | no |
+| [5 - Recommended](#Recommended) | The feature/resource is recommended for general use. | yes |
```
I don't see much benefit between Ready for Release, Released, and Recommended, tbh. Maybe I'm thinking too narrowly about only code things, but it seems that if a feature is shipped, no one is really going to check the RFC stage to see whether they should use it. A beta warning in JS console would go much further.

Another way to model this (again, limited to code things) is to implement in userland (via addon) and then incorporate into "core" to signal that it's recommended. I just can't see myself checking RFCs other than to check if the feature is released, and in what version.

kategengler replied:
"Recommended" now has more detail. 

"Ready for Release" gives us a point at which all the teams and contributors can confirm the work is done and ready to be "Released" with stability guarantees. Once something is "Released", it's the same as it is now, it is stable. I don't think we'd ever expect the average user to check the RFC stage to decide whether to use a feature, but the stage will inform how the community talks about or promotes a feature. It's possible we could expose the stage in a user-friendly format in a dashboard.

For example, "Template-only components" could reach "Released" and be usable, with API documentation, but take some time to reach the point where the codemods were hardened, the guides and tutorial had all been updated to use them where appropriate. At that point, we have another PR & FCP for the community and teams to agree that the feature is "Recommended"

rwjblue commented:
We discussed this in the most recent core team meeting, and are moving it into final comment period.

rwjblue commented:
OK, lets do it!!!!

jelhan commented on text/0617-rfc-stages.md:
```diff
@@ -0,0 +1,409 @@
+
+#### Discontinued
+
+An previously [Accepted](#Accepted) RFC may be discontinued at any point. The RFC
```
Neither _closed_ nor _discontinued_ seem to apply for an RFC, which is in _exploring_ stage. Did I missed something?

kategengler replied:
Good catch -- it was intended that Exploring RFCs can move to closed via FCP, same as proposed. I'll update.
